packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,519 @@
+    const foundValue = await storageTrie.get(key)
+    return Buffer.from(RLP.decode(Uint8Array.from(foundValue ?? [])) as Uint8Array)
+  }
+
+  /**
+   * Retrieves a storage slot from the provider and stores the proof nodes in the local trie
+   * @param address - Address to be retrieved from provider
+   * @param key - Key of storage slot to be returned
+   * @private
+   */
+  private async getContractStorageFromProvider(address: Address, key: Buffer): Promise<void> {
+    if (this.externallyRetrievedStorageKeys.has(address.toString())) {
+      const map = this.externallyRetrievedStorageKeys.get(address.toString())
+      if (map?.get(key.toString('hex')) !== undefined) {
+        // Return early if slot has already been retrieved from provider
+        return
+      }
+    }
+
+    const accountData = await this.provider.send('eth_getProof', [
+      address.toString(),
+      [bufferToHex(key)],
+      this.blockTag,
+    ])
+
+    const rawAccountProofData = accountData.accountProof
+    await this.trie.fromProof(rawAccountProofData.map((e: string) => hexToBytes(e)))
+
+    const storageData = accountData.storageProof.find(
+      (el: any) => el.key === '0x' + key.toString('hex')
+    )
+
+    const storageTrie = await this._getStorageTrie(address)
+
+    await storageTrie.fromProof(storageData.proof.map((e: string) => hexToBytes(e)))
+
+    let map = this.externallyRetrievedStorageKeys.get(address.toString())
+    if (!map) {
+      this.externallyRetrievedStorageKeys.set(address.toString(), new Map())
+      map = this.externallyRetrievedStorageKeys.get(address.toString())!
+    }
+    map.set(key.toString('hex'), true)
+  }
+
+  /**
+   * Adds value to the state trie for the `account`
+   * corresponding to `address` at the provided `key`.
+   * @param address - Address to set a storage value for
+   * @param key - Key to set the value at. Must be 32 bytes long.
+   * @param value - Value to set at `key` for account corresponding to `address`.
+   * Cannot be more than 32 bytes. Leading zeros are stripped.
+   * If it is empty or filled with zeros, deletes the value.
+   */
+  async putContractStorage(address: Address, key: Buffer, value: Buffer): Promise<void> {
+    await this.getContractStorageFromProvider(address, key)
+    const storageTrie = await this._getStorageTrie(address)
+
+    if (key.length !== 32) {
+      throw new Error('Storage key must be 32 bytes long')
+    }
+
+    if (value.length > 32) {
+      throw new Error('Storage value cannot be longer than 32 bytes')
+    }
+
+    value = unpadBuffer(value)
+    const encodedValue = Buffer.from(RLP.encode(Uint8Array.from(value)))
+    if (value.length > 0) {
+      await storageTrie.put(key, encodedValue)
+    } else {
+      try {
+        await storageTrie.del(key)
+      } catch (err: any) {
+        if (
+          err.message === 'Missing node in DB' &&
+          (err.stack as string).includes('async Trie.del')
+        ) {
+          throw new Error(
```

Comment:
can we also attach the original stack of the `err` here as it would be informational